File: examples/cpp/04_reduction/main.cpp
#include <iostream>

#include <occa.hpp>

//---[ Internal Tools ]-----------------
// Note: These headers are not officially supported
//       Please don't rely on it outside of the occa examples
#include <occa/internal/utils/cli.hpp>
#include <occa/internal/utils/testing.hpp>
//======================================

occa::json parseArgs(int argc, const char **argv);

int main(int argc, const char **argv) {
  occa::json args = parseArgs(argc, argv);

  occa::setDevice((std::string) args["options/device"]);

  // Choosing something not divisible by 256
  int entries = 10000;
  int block   = 256;
  int blocks  = (entries + block - 1)/block;

  float *vec      = new float[entries];
  float *blockSum = new float[blocks];
  float atomicSum = 0;

  float sum = 0;

  // Initialize device memory
  for (int i = 0; i < entries; ++i) {
    vec[i] = 1;
    sum   += vec[i];
  }

  for (int i = 0; i < blocks; ++i) {
    blockSum[i] = 0;
  }

  // Allocate memory on the device
  occa::memory o_vec       = occa::malloc<float>(entries);
  occa::memory o_blockSum  = occa::malloc<float>(blocks);
  occa::memory o_atomicSum = occa::malloc<float>(1, &atomicSum);

  // Pass value of 'block' at kernel compile-time
  occa::json reductionProps({
    {"defines/block", block},
  });

  occa::kernel reductionWithSharedMemory = (
    occa::buildKernel("reduction.okl",
                      "reductionWithSharedMemory",
                      reductionProps)
  );

  occa::kernel reductionWithAtomics = (
    occa::buildKernel("reduction.okl",
                      "reductionWithAtomics",
                      reductionProps)
  );

  // Host -> Device
  o_vec.copyFrom(vec);

  reductionWithSharedMemory(entries, o_vec, o_blockSum);
  reductionWithAtomics(entries, o_vec, o_atomicSum);

  // Host <- Device
  o_blockSum.copyTo(blockSum);
  o_atomicSum.copyTo(&atomicSum);

  // Finalize the reduction in the host
  for (int i = 1; i < blocks; ++i) {
    blockSum[0] += blockSum[i];
  }

  // Validate
  bool hasError = false;
  if (!occa::areBitwiseEqual(blockSum[0], sum)) {
    std::cout << "sum      = " << sum << '\n'
              << "blockSum = " << blockSum[0] << '\n';

    std::cout << "(Shared) Reduction failed\n";
    hasError = true;
  }
  else {
    std::cout << "(Shared) Reduction = " << blockSum[0] << '\n';
  }

  if (!occa::areBitwiseEqual(atomicSum, sum)) {
    std::cout << "sum       = " << sum << '\n'
              << "atomicSum = " << atomicSum << '\n';

    std::cout << "(Atomic) Reduction failed\n";
    hasError = true;
  }
  else {
    std::cout << "(Atomic) Reduction = " << blockSum[0] << '\n';
  }

  if (hasError) {
    throw 1;
  }

  // Free host memory
  delete [] vec;
  delete [] blockSum;

  // Device memory is automatically freed

  return 0;
}

occa::json parseArgs(int argc, const char **argv) {
  occa::cli::parser parser;
  parser
    .withDescription(
      "Example of a reduction kernel which sums a vector in parallel"
    )
    .addOption(
      occa::cli::option('d', "device",
                        "Device properties (default: \"{mode: 'Serial'}\")")
      .withArg()
      .withDefaultValue("{mode: 'Serial'}")
    )
    .addOption(
      occa::cli::option('v', "verbose",
                        "Compile kernels in verbose mode")
    );

  occa::json args = parser.parseArgs(argc, argv);
  occa::settings()["kernel/verbose"] = args["options/verbose"];

  return args;
}

File: examples/fortran/04_reduction/main.f90
program main
  use occa

  implicit none

  integer :: alloc_err, i
  integer(occaUDim_t), parameter :: entries = 10000
  integer(occaUDim_t), parameter :: blk = 256
  integer(occaUDim_t), parameter :: blks = (entries + blk - 1)/blk
  character(len=1024) :: arg, info

  real(C_float), allocatable, target :: vec(:), blockSum(:)
  real(C_float) :: sig

  ! OCCA device, kernel, memory and property objects
  type(occaKernel) :: reduction
  type(occaMemory) :: o_vec, o_blockSum
  type(occaJson)   :: reductionProps

  ! Set default OCCA device info
  info = "{mode: 'Serial'}"
  !info = "{mode: 'OpenMP', schedule: 'compact', chunk: 10}"
  !info = "{mode: 'CUDA'  , device_id: 0}"
  !info = "{mode: 'OpenCL', platform_id: 0, device_id: 0}"

  ! Parse command arguments
  i = 1
  do while (i .le. command_argument_count())
    call get_command_argument(i, arg)

    select case (arg)
      case ("-v", "--verbose")
        call occaJsonObjectSet(occaSettings(), "kernel/verbose", occaTrue)
      case ("-d", "--device")
        i = i+1
        call get_command_argument(i, info)
      case ("-h", "--help")
        call print_help()
        stop
      case default
        write(*,'(2a, /)') "Unrecognised command-line option: ", arg
        stop
    end select
    i = i+1
  end do

  ! Allocate host memory
  allocate(vec(1:entries), blockSum(1:blks), stat=alloc_err)
  if (alloc_err /= 0) stop "*** Not enough memory ***"

  ! Initialize host memory
  vec = 1
  blockSum = 0
  sig = sum(vec)

  ! Print device infos
  call occaPrintModeInfo()

  ! Create OCCA device
  call occaSetDeviceFromString(F_C_str(info))

  ! Allocate memory on the device
  o_vec      = occaTypedMalloc(entries, occaDtypeFloat, C_NULL_ptr, occaDefault)
  o_blockSum = occaTypedMalloc(blks,    occaDtypeFloat, C_NULL_ptr, occaDefault)

  ! Pass value of 'block' at kernel compile-time
  reductionProps = occaCreateJson()
  call occaJsonObjectSet(reductionProps, F_C_str("defines/block"), occaInt(blk))

  reduction = occaBuildKernel(F_C_str("reduction.okl"), &
                              F_C_str("reduction"), &
                              reductionProps)

  ! Host -> Device
  call occaCopyPtrToMem(o_vec, C_loc(vec), occaAllBytes, 0_occaUDim_t, occaDefault)

  call occaKernelRun(reduction, occaInt(entries), o_vec, o_blockSum)

  ! Host <- Device
  call occaCopyMemToPtr(C_loc(blockSum), o_blockSum, occaAllBytes, 0_occaUDim_t, occaDefault)

  blockSum(1) = sum(blockSum)

  ! Validate
  if (abs(blockSum(1) - sig) > 1.0e-8) then
    write(*,*) "sum      = ", sig
    write(*,*) "blockSum = ", blockSum(1)
    stop "*** Reduction failed ***"
  else
    write(*,*) "Reduction = ", blockSum(1)
  end if

  ! Free host memory
  deallocate(vec, blockSum, stat=alloc_err)
  if (alloc_err /= 0) stop "*** Deallocation not successful ***"

  ! Free device memory and OCCA objects
  call occaFree(reductionProps)
  call occaFree(reduction)
  call occaFree(o_vec)
  call occaFree(o_blockSum)

contains
  subroutine print_help()
    write(*,'(a, /)') "Example of a reduction kernel which sums a vector in parallel"
    write(*,'(a, /)') "command-line options:"
    write(*,'(a)')    "  -v, --verbose     Compile kernels in verbose mode"
    write(*,'(a)')    "  -d, --device      Device properties (default: ""{mode: 'Serial'}"")"
    write(*,'(a)')    "  -h, --help        Print this information and exit"
  end subroutine print_help
end program main
